### PR TITLE
Seed order to manage logs overflow

### PR DESCRIPTION
**Pull request by amirylm**
Using deterministic seed ordering to drop the same logs across the network in cases of logs overflow, where we hit the limits:

- `buffer.enqueue()` - ensures that buffer size won't exceed limits, while dropping the same logs across the network
- `provider.GetLatestPayloads()` - ensures that in cases of multiple results, the same set of logs are used across the network
- `recoverer.GetRecoveryProposals()` - same as latest payloads

### CODE REVIEW

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go [hunk 1]
```diff
@@ -318,7 +318,7 @@ func (p *logEventProvider) updateFiltersLastPoll(entries []upkeepFilter) {
 	p.filterStore.UpdateFilters(func(orig, f upkeepFilter) upkeepFilter {
 		if f.lastPollBlock > orig.lastPollBlock {
 			orig.lastPollBlock = f.lastPollBlock
-			p.lggr.Debugw("Updated lastPollBlock", "lastPollBlock", f.lastPollBlock, "upkeepID", f.upkeepID)
+			// p.lggr.Debugw("Updated lastPollBlock", "lastPollBlock", f.lastPollBlock, "upkeepID", f.upkeepID)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/provider.go:321]: this log is pretty useful to debug progress (and helped catch a bug). Maybe we can log this on some sample (like 1%)?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 2]
```diff
@@ -596,3 +604,18 @@ func (r *logRecoverer) removePending(workID string) {
 	}
 	r.pending = updated
 }
+
+// sortPending sorts the pending list by a random order based on the latest block.
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:608]: this will be different order on nodes as latestBlock will be out of sync. We don't really need consensus in recoverables so it should be ok here, but in provider we can use `latestBlock % 100` maybe to ensure it stays same across nodes for some time

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
+	"encoding/hex"
 	"math/big"
 	"sort"
 	"sync"
 	"sync/atomic"
 
 	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/ocr2keepers/pkg/v3/random"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg/v3/types"
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:13]: not fully related to this PR, but can we clean up the variables in line 17-20

allowedLogsPerBlock = 128 -> maxLogsPerUpkeepPerBlock, set it to something like 10?
bufferMaxBlockSize = 1024 -> maxLogsPerBlock

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go [hunk 4]
```diff
@@ -33,9 +36,46 @@ type fetchedBlock struct {
...
+		for _, l := range currentBlock.logs {
+			if dropped.upkeepID == nil && l.upkeepID.Cmp(fl.upkeepID) == 0 {
+				dropped = l
+				continue
+			}
+			currentLogs = append(currentLogs, l)
+		}
+		currentBlock.logs = currentLogs
+		return dropped, true
+	} else if len(currentBlock.logs) > maxBlockLogs {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:64]: should this be `len(currentBlock.logs)` + `len(currentBlock.visited)` ? (We account for both in per upkeep limit)

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go (hunk 2 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go]: Used `latestBlock / 10`, do you think we need 100? (maybe for fast chains)

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go (hunk 3 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:13]: I've set `maxLogsPerUpkeepInBlock` to 32

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go [hunk 5]
```diff
@@ -62,6 +103,27 @@ func (b fetchedBlock) Clone() fetchedBlock {
...
+			Index:     uint32(log.log.LogIndex),
+			BlockHash: log.log.BlockHash,
+		}
+		logID := hex.EncodeToString(ext.LogIdentifier())
+		shuffledLogIDs[i] = random.ShuffleString(logID, randSeed)
+	}
+
+	sort.Slice(b.logs, func(i, j int) bool {
+		return shuffledLogIDs[i] < shuffledLogIDs[j]
+	})
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:124]: just want to verify this will work properly if the indices passed in comparator doesn't refer to the slice being sorted (instead a constant slice)?

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go [hunk 6]
```diff
@@ -596,3 +619,23 @@ func (r *logRecoverer) removePending(workID string) {
 	}
 	r.pending = updated
 }
+
+// sortPending sorts the pending list by a random order based on the normalized latest block number.
+// Divided by 10 to ensure that nodes with similar block numbers won't end up with different order.
+// NOTE: the lock must be held before calling this function.
+func (r *logRecoverer) sortPending(latestBlock uint64) {
+	normalized := latestBlock / 10
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:627]: nit. 10 might not be enough on fast chains like arb. 100 might be better (but then its slow on other chains also )

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 7]
```diff
@@ -283,6 +283,7 @@ func (r *EvmRegistry) refreshActiveUpkeeps() error {
 		switch core.GetUpkeepType(*uid) {
 		case ocr2keepers.LogTrigger:
 			logTriggerIDs = append(logTriggerIDs, id)
+		default:
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:286]: is this related to this PR?

**Review by infiloop2 — DISMISSED** [review]
overall looks good! just couple of non blocking nits

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go (hunk 5 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:124]: Yes, the comparator accepts indices and returns a bool, making it completely decoupled from the underlying slice.

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 7 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:286]: No but that's a lint issue :)

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go (hunk 6 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:627]: We know the block time within the recoverer, we could also compute something based on the block time.

Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/recoverer.go:627]: that can have further issues when blockTime is not synchronized 😅

File: core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go [hunk 8]
```diff
@@ -62,6 +118,22 @@ func (b fetchedBlock) Clone() fetchedBlock {
...
 
+// Sort by log identifiers, shuffled using a pseduorandom souce that is shared across all nodes
+// for a given block.
+func (b *fetchedBlock) Sort() {
+	randSeed := random.GetRandomKeySource(nil, uint64(b.blockNumber))
+
+	shuffledLogIDs := make(map[string]string, len(b.logs))
+	for _, log := range b.logs {
+		logID := log.getLogID()
+		shuffledLogIDs[logID] = random.ShuffleString(logID, randSeed)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/logprovider/buffer.go:129]: nit. i think this will always be constant for a given log, so we can also make it a part of logID itself

**Review by infiloop2 — APPROVED** [review]
